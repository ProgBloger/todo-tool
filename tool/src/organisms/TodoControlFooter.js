import style from "./TodoControlFooter.module.scss";
import React from "react";

const TodoControlFooter = () => {

    return (
        <React.Fragment>
            <div className={style.container}>
                <span className={style.left}>
                    <label>{5} items left</label>
                </span>
                <span className={style.center}>
                    <button>All</button>
                    <button>Active</button>
                    <button>Completed</button>
                </span>
                <span className={style.right}>
                    <button>Clear Completed</button>
                </span>
            </div>
        </React.Fragment>
    )
}

export default TodoControlFooter;