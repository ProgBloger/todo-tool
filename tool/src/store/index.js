import {combineReducers, configureStore} from "@reduxjs/toolkit";
import todoListSlice from "./todo-list-slice";

var persistedState = JSON.parse(localStorage.getItem("state"));

const store = configureStore({
    reducer: {
        todoList: todoListSlice.reducer,
    },
    preloadedState: persistedState === null ? {} : persistedState
});

store.subscribe(() => {
    var state = store.getState();

    var jsonState = JSON.stringify(state);
    localStorage.setItem("state", jsonState);
});

export default store;

export const setupStore = preloadedState => {
    return configureStore({
        reducer: {
            todoList: todoListSlice.reducer,
        },
        preloadedState: preloadedState
    })
}