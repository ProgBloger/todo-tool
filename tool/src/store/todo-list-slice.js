import { createSlice } from "@reduxjs/toolkit";

const initialState = {
    items: [],
    itemsToShow:[],
    itemsTotal: 0,
    itemsTotalActive: 0,
    // true - completed
    // false - active
    // null - all
    filter: null,
};

const filterProcessor = (itemsList, filterParam) => {
    if(filterParam == null)
        return itemsList;
    return itemsList.filter(item => item.completed == filterParam)
}

const todoListSlice = createSlice({
    name: 'todoList',
    initialState,
    reducers: {
        addItem(state, action) {
            state.itemsTotal = state.itemsTotal + 1;
            state.itemsTotalActive = state.itemsTotalActive + 1;
            console.log(action);
            const newItem = {
                id: state.itemsTotal,
                text: action.payload.text,
                completed: false,
            }

            state.items.push(newItem);

            // update the show list based on active filter
            state.itemsToShow = filterProcessor(state.items, state.filter);
        },
        removeItem(state, action) {
            const id = action.payload.id;
            const existingItem = state.items.find(item => item.id === id);
            
            // updating totals on item being removed
            state.itemsTotal--;
            if(!existingItem.completed) {
                state.itemsTotalActive--;
            }

            state.items = state.items.filter(item => item.id !== id);

            // update the show list based on active filter
            state.itemsToShow = filterProcessor(state.items, state.filter);
        },
        changeItemState(state, action) {
            // find item from the state list
            const newItemState = action.payload;
            const foundItem = state.items.find((item) => item.id === newItemState.id);
            
            // set item's state from the payload
            foundItem.completed = newItemState.completed;
            
            if(foundItem.completed) {
                state.itemsTotalActive--;
            } else {
                state.itemsTotalActive++;
            }
            
            // update the show list based on active filter
            state.itemsToShow = filterProcessor(state.items, state.filter);
        },
        fiterItems(state, action) {
            state.filter = action.payload
            state.itemsToShow = filterProcessor(state.items, state.filter);
        },
        clearCompleted(state, action){},
    },
});

export const todoActions = todoListSlice.actions;
export default todoListSlice;

