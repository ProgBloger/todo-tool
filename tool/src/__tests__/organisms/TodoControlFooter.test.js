import renderer, { act } from 'react-test-renderer';
import expect from 'expect';
import TodoControlFooter from '../../organisms/TodoControlFooter';
import React from 'react';
import store, { setupStore } from '../../store';
import { Provider } from 'react-redux';
import { renderWithProviders } from '../../utils/test-utils';
import { screen } from '@testing-library/react';
import { todoActions } from '../../store/todo-list-slice';

it('renders', () => {
    // Arrange
    // Act
    var tree = renderer.create(
        <Provider store={store}>
        <TodoControlFooter />
        </Provider>
    ).toJSON();

    // Assert
    expect(tree).toMatchSnapshot();
})



it('if task change to complete items left decremented', () => {
    // Arrange
    const initialItem = { 
        id: 0, 
        text: 'Pet a cat', 
        completed: false
    };

    const {addItem, changeItemState} = todoActions;
    
    const store = setupStore();
    store.dispatch(addItem(initialItem));
    renderWithProviders(
            <TodoControlFooter />, {
                store: store,
            });
    
    const itemId = store.getState().todoList.items[0].id;
    
    // Act
    act(() => {
        store.dispatch(changeItemState({
            id: itemId, 
            completed: !initialItem.completed,}));
    });

    // Assert
    const expectedLabel = screen.getByText(/0 items left/i);

    let state = store.getState();
    expect(state.todoList.itemsTotalActive).toBe(0);

    expect(expectedLabel)
        .toBeDefined();
})

// if task change to active items left incremented
// if task added items left incremented

// if All clicked all tasks are in show array
// if All clicked All button has activeFilter class

// if Active clicked only active tasks are in show array
// if Active clicked All button has activeFilter class

// if Completed clicked all tasks are in show array
// if Completed clicked All button has activeFilter class

// if Clear Completed clicked no Complete items present