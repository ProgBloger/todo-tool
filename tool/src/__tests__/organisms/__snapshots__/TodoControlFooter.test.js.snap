// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<div
  className="container"
>
  <span
    className="left"
  >
    <label>
      0
       items left
    </label>
  </span>
  <span
    className="center"
  >
    <button
      className="activeFilter"
      onClick={[Function]}
    >
      All
    </button>
    <button
      onClick={[Function]}
    >
      Active
    </button>
    <button
      onClick={[Function]}
    >
      Completed
    </button>
  </span>
  <span
    className="right"
  >
    <button
      onClick={[Function]}
    >
      Clear Completed
    </button>
  </span>
</div>
`;
